Make StationLookup a proper Mapping

dopplershift opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/Unidata/MetPy/blob/main/CONTRIBUTING.md
-->

#### Description Of Changes
This allows iteration, membership, etc. directly on the station information rather than needing to use the `tables` property. The abstract base class gives us some things like `get()`, `keys()`, and `values()` for free.

<!--
Feel free to remove check-list items aren't relevant to your change

Please use keywords (e.g., Fixes, Closes) to create link to the issues or pull
requests you resolved, so that they will automatically be closed when your pull
request is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->

#### Checklist

- [x] Addressing some of #2397 
- [x] Fixes #2343
- [x] Tests added
- [x] Fully documented

dopplershift commented:
@jthielen Do you think it would be enough to have a docstring example of use here? Of course, `station_info` and `StationLookup` don't seem to show up in our reference pages atm. Or did you have a more proper gallery example in mind?

jthielen commented:
> @jthielen Do you think it would be enough to have a docstring example of use here? Of course, `station_info` and `StationLookup` don't seem to show up in our reference pages atm. Or did you have a more proper gallery example in mind?

Yes and yes? Having `StationLookup` in the API reference with a docstring example (and description of where this info is from) was my initial thought, but having a gallery example would be nice as well. Given that we have geopandas in the docs, for the latter I was thinking a demo loading the station info into a GeoDataFrame and then applying some example spatial operation? I should be able to put something like that together quick tomorrow.

dopplershift commented:
Ok, I added `StationLookup` and `station_info` to the docs for `metpy.io`, and I added a brief docstring example.

I'll leave a proper example to later, when I can come up with something useful--or if you get some cycles to breathe life into your idea @jthielen .

dcamron commented on src/metpy/io/station_data.py:
```diff
@@ -120,8 +121,17 @@ def _read_airports_file(input_file=None):
+    """Look up station information from multiple sources.
+
+    This class follows the `Mapping` protocol with station ID as the key. This makes it
+    possible to e.g. iterate over all locations and get all of a certain criteria:
+    >>> import metpy.io
```
```suggestion
    possible to e.g. iterate over all locations and get all of a certain criteria:
    
    >>> import metpy.io
```
Else the code sample doesn't render and format appropriately.

dopplershift replied:
Good catch.

dcamron reviewed (changes requested):
Looks nice and helpful! Small change for docs render. Also, question: are instances exported with the `Exporter` context manager supposed to automatically end up with generated doc summaries? If so, I'm not seeing that here after a quick glance.

dopplershift commented:
`Exporter`'s primary mission in life is managing `__all__`. Sphinx doesn't appear to do anything with instances, and so far I haven't found any magic incantations to change that. I suppose we could implement a custom module template. 😬

dcamron commented:
I think I'm very okay skipping that and calling the explicit shoutout in the class docstring good enough!
